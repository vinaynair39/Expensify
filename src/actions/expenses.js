import  database  from '../firebase/firebase';

export const addExpense = (expense) => ({
    type: "ADD_EXPENSE",
    expense
  });

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const{
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;
    
    const expense = {description,amount,note,createdAt};
    return database.ref('expenses').push(expense).then((ref)=> {
      console.log(ref.key);
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  };
};

export const editExpense = (id, updates = {}) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});

export const setEditExpense = (id, updates ={}) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).update(updates).then((ref)=> {
      dispatch(editExpense(id, updates));
    });
  };
};
  export const removeExpense = (id) => ({
    type: "REMOVE_EXPENSE",
    id
  });
  


  
export const setExpenses = (expenses) => {
  return {
    type: "SET_EXPENSES",
    expenses
  };
}

export const startSetExpenses = () => {
  return (dispatch) => {
    return database.ref('expenses').once('value').then((snapshot) => {
      const expenses = [];
      snapshot.forEach((childSnapshot) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });
      dispatch(setExpenses(expenses));
    })
    }
  }

export const setRemoveExpense = (id) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove().then(() => {
      dispatch(removeExpense(id));
    });
  }
}