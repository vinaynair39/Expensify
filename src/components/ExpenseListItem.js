import React from 'react';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import moment from 'moment';

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <div>
    <Link to={`/edit/${id}`}>
      <h3>{description}</h3>
    </Link>
    <p>{numeral(amount).format('0,0')} - {moment(createdAt).format("Do MMM YYYY")}</p>
  </div>
);


export default ExpenseListItem;
