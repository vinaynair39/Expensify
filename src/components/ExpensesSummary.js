import React from 'react';
import getVisibleExpenses from '../selectors/expenses';
import getTotalAmount from '../selectors/expenses-total';
import {connect} from 'react-redux';
import numeral from 'numeral';


export const ExpensesSummary = (props) => {
    const word = props.expenses.length <= 1 ?  'expense' : 'expenses';
    return(
        <h3>Viewing {props.expenses.length} {word} totaling: {numeral(getTotalAmount(props.expenses)).format('0,0')}</h3>
    );
}

const mapStateToProps = (state) => {
    return{
        expenses: getVisibleExpenses(state.expenses, state.filters),
    }
}

export default connect(mapStateToProps)(ExpensesSummary);