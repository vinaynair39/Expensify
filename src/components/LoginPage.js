import React from 'react';
import {connect} from 'react-redux';
import {startLogin} from '../actions/auth';

export const LoginPage = (props) => {
    return(
        <div>
            <h1>
                <button onClick = {props.Login}>Login</button>
                <button>Sign Up!</button>
            </h1>
        </div>
    );
};

const mapDispatchToProps = (dispatch) => {
    return {
        Login : () => dispatch(startLogin()),
    };
};

export default connect(undefined, mapDispatchToProps)(LoginPage);