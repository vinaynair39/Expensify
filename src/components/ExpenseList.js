import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';
import getTotalAmount from '../selectors/expenses-total';

export const ExpenseList = (props) => (
  <div>
    {props.expenses.length === 0 ?   (<p>no expense</p>) : props.expenses.map((expense) => {
      return <ExpenseListItem key={expense.id} {...expense} />;
    })}
  
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
