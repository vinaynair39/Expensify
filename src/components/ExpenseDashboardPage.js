import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';
import {NavLink} from 'react-router-dom';

const ExpenseDashboardPage = () => (
  <div>
    <NavLink to="/dashboard" activeClassName="is-active" exact={true}>Dashboard</NavLink>
    <NavLink to="/create" activeClassName="is-active">Create Expense</NavLink>
    <ExpenseListFilters />
    <ExpensesSummary/>
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;
