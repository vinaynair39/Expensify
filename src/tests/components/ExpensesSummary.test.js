import React from 'react';
import {ExpensesSummary} from '../../components/ExpensesSummary';
import {shallow} from 'enzyme';
import expenses from '../fixtures/expenses';


test('Should render ExpensesSummary with 1 expense', () => {
    const wrapper = shallow(<ExpensesSummary expenses={[expenses[2]]}/>);
    expect(wrapper).toMatchSnapshot();
});

test('Should render ExpensesSummary with multiple expense', () => {
    const wrapper = shallow(<ExpensesSummary expenses={expenses}/>);
    expect(wrapper).toMatchSnapshot();
});