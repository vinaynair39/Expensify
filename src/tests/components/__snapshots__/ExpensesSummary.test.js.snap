// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpensesSummary with 1 expense 1`] = `
<h3>
  Viewing 
  1
   
  expense
   totaling: 
  4,500
</h3>
`;

exports[`Should render ExpensesSummary with multiple expense 1`] = `
<h3>
  Viewing 
  3
   
  expenses
   totaling: 
  114,195
</h3>
`;
